Moves intern and label modules into bevy_ecs

mnmaita opened the pull request:
# Objective

- Attempts to solve two items from https://github.com/bevyengine/bevy/issues/11478.

## Solution

- Moved `intern` module from `bevy_utils` into `bevy_ecs` crate and updated all relevant imports.
- Moved `label` module from `bevy_utils` into `bevy_ecs` crate and updated all relevant imports.

---

## Migration Guide

- Replace `bevy_utils::define_label` imports with `bevy_ecs::define_label` imports.
- Replace `bevy_utils::label::DynEq` imports with `bevy_ecs::label::DynEq` imports.
- Replace `bevy_utils::label::DynHash` imports with `bevy_ecs::label::DynHash` imports.
- Replace `bevy_utils::intern::Interned` imports with `bevy_ecs::intern::Interned` imports.
- Replace `bevy_utils::intern::Internable` imports with `bevy_ecs::intern::Internable` imports.
- Replace `bevy_utils::intern::Interner` imports with `bevy_ecs::intern::Interner` imports.

james7132 reviewed (commented):
Mind fixing CI? The lints seem like they should be fixable by using `cast::<()>` instead of `as`.

mnmaita commented:
> Mind fixing CI? The lints seem like they should be fixable by using `cast::<()>` instead of `as`.

Done, rebased and updated the migration guide. Feel free to take a second look at it now @james7132. Thanks!

mnmaita commented on crates/bevy_ecs/src/label.rs:
```diff
@@ -125,7 +125,7 @@ macro_rules! define_label {
             fn dyn_hash(&self, state: &mut dyn ::std::hash::Hasher);
 
-            /// Returns an [`Interned`](bevy_utils::intern::Interned) value corresponding to `self`.
+            /// Returns an [`Interned`](bevy_ecs::intern::Interned) value corresponding to `self`.
```
Changing this to `[`Interned`](Interned)` fixes one of the issues in the docs CI job but there seems to be another issue that I couldn't figure out yet. I'd appreciate some guidance with it.

james7132 replied:
```suggestion
            /// Returns an [`Interned`](crate::intern::Interned) value corresponding to `self`.
```

`crate` is self-referential when inside the target crate.

mnmaita replied:
Thanks for applying that change. I'm 99% positive that I tried that and it error-ed out locally but I might be wrong. Let's see if that fixes the CI issue.

mnmaita commented:
@james7132 I did a change that fixed the issue with `Interned` in the docs, but now there's a different one... Could you please help me figure out why `Schedule` is giving errors during doc generation? I don't really know what's the cause of this issue. Thanks!

james7132 commented on crates/bevy_ecs/src/schedule/set.rs:
```diff
@@ -3,14 +3,16 @@ use std::fmt::Debug;
+    },
 };
 
 define_label!(
```
I think you just need to make this ``[`Schedule`](crate::schedule::Schedule)``.

mnmaita replied:
This should be fixed in the latest change @james7132
